posthog/api/query_ws.py

```diff
@@ -0,0 +1,115 @@
+from posthog.api.services.query import process_query_model
+from posthog.hogql_queries.query_runner import ExecutionMode, execution_mode_from_refresh
+from posthog.clickhouse.query_tagging import tag_queries
+from pydantic import BaseModel
+from posthog.errors import ExposedCHQueryError
+from posthog.hogql.errors import ExposedHogQLError
+from sentry_sdk import capture_exception
+from asgiref.sync import sync_to_async
+
+from posthog.rate_limit import ClickHouseBurstRateThrottle, ClickHouseSustainedRateThrottle, HogQLQueryThrottle
+from posthog.api.websocket import BaseWebsocketConsumer
+import json
+import uuid
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+
+class Response(BaseModel, Generic[T]):
+    status: int
+    result: T
+    client_query_id: str
+
+
+def _query(user, query, execution_mode, query_id):
+    return process_query_model(
+        user.team,
+        query,
+        execution_mode=execution_mode,
+        query_id=query_id,
+        user=user,
+    )
+
+
+class QueryConsumer(BaseWebsocketConsumer):
+    def get_throttles(self, data):
+        if query := data.get("query"):
+            if isinstance(query, dict) and query.get("kind") == "HogQLQuery":
+                return [HogQLQueryThrottle()]
+        return [ClickHouseBurstRateThrottle(), ClickHouseSustainedRateThrottle()]
+
+    async def connect(self):
+        # Accept the WebSocket connection
+        if not await self.check_authentication():
+            return
+        await self.accept()
+
+    async def disconnect(self, close_code):
+        # Handle WebSocket disconnection (if needed)
+        pass
```

Comment:
Wouldn't we want to cancel any pending HogQL queries here?

Reply:
This may be tricky given that `process_query_model` is not a coroutine, so it will not handle any cancellation internally.

Reply:
Unless that is already happening within `process_query_model`. If so, then please ignore this.

Reply:
Not totally, the browser can disconnect when a user switches tabs (ie when they're waiting on a long running query to complete) so we don't want to kill a query then. We already have the kill_query mechanism on the frontend when a user navigates away, and that should also kill these queries.